artemis-protocols/artemis-openwire-protocol/src/main/java/org/apache/activemq/artemis/core/protocol/openwire/OpenWireMessageConverter.java

```diff
@@ -128,389 +110,159 @@ public Object outbound(org.apache.activemq.artemis.api.core.Message message, int
-               break;
-            case org.apache.activemq.artemis.api.core.Message.MAP_TYPE:
-               writeMapType(contents, messageCompressed, body);
-               break;
-            case org.apache.activemq.artemis.api.core.Message.OBJECT_TYPE:
-               writeObjectType(contents, messageCompressed, body);
-               break;
-            case org.apache.activemq.artemis.api.core.Message.STREAM_TYPE:
-               writeStreamType(contents, messageCompressed, body);
-               break;
-            case org.apache.activemq.artemis.api.core.Message.BYTES_TYPE:
-               writeBytesType(contents, messageCompressed, body);
-               break;
-            default:
-               writeDefaultType(contents, messageCompressed, body);
-               break;
-         }
+      final ActiveMQBuffer body = openwireMessage.getBodyBuffer();
+      final boolean messageCompressed = messageSend.isCompressed();
+      if (messageCompressed) {
+         openwireMessage.putBooleanProperty(AMQ_MSG_COMPRESSED, messageCompressed);
       }
+      final ByteSequence contents = messageSend.getContent();
+      OpenWireCoreConverter.writeContentIntoBody(body, contents, coreType, messageCompressed);
       //amq specific
-      coreMessage.putLongProperty(AMQ_MSG_ARRIVAL, messageSend.getArrival());
-      coreMessage.putLongProperty(AMQ_MSG_BROKER_IN_TIME, messageSend.getBrokerInTime());
+      openwireMessage.putLongProperty(AMQ_MSG_ARRIVAL, messageSend.getArrival());
+      openwireMessage.putLongProperty(AMQ_MSG_BROKER_IN_TIME, messageSend.getBrokerInTime());
       final BrokerId[] brokers = messageSend.getBrokerPath();
       if (brokers != null) {
-         putMsgBrokerPath(brokers, coreMessage);
+         putMsgBrokerPath(brokers, openwireMessage);
       }
       final BrokerId[] cluster = messageSend.getCluster();
       if (cluster != null) {
-         putMsgCluster(cluster, coreMessage);
+         putMsgCluster(cluster, openwireMessage);
       }
 
-      coreMessage.putIntProperty(AMQ_MSG_COMMAND_ID, messageSend.getCommandId());
+      openwireMessage.putIntProperty(AMQ_MSG_COMMAND_ID, messageSend.getCommandId());
       final String corrId = messageSend.getCorrelationId();
       if (corrId != null) {
-         coreMessage.putStringProperty(JMS_CORRELATION_ID_PROPERTY, new SimpleString(corrId));
+         openwireMessage.putStringProperty(JMS_CORRELATION_ID_PROPERTY, new SimpleString(corrId));
       }
       final DataStructure ds = messageSend.getDataStructure();
       if (ds != null) {
-         putMsgDataStructure(ds, marshaller, coreMessage);
+         putMsgDataStructure(ds, marshaller, openwireMessage);
       }
       final String groupId = messageSend.getGroupID();
       if (groupId != null) {
-         coreMessage.putStringProperty(AMQ_MSG_GROUP_ID, new SimpleString(groupId));
+         openwireMessage.putStringProperty(AMQ_MSG_GROUP_ID, new SimpleString(groupId));
```

Comment:
this could be pooled with `coreMessageObjectPools` as mentioned above

Reply:
I assume you meant *should

Reply:
as per other comments, id prefer we merge this, and then iterate over the next week, with smaller PR's with these optimisations.